metrics/server.go

```diff
@@ -0,0 +1,384 @@
+
+			smT, err := calculateSummarizedMetric(metrics.MetricsOverTransactions)
+			if err != nil {
+				log.Logger.Debugf("unable to compute summaried metrics over transactions for application %s (version %d; signature %s)", application, v, *signature)
+				continue
+			} else {
+				entry.SummaryOverTransactions = append(entry.SummaryOverTransactions, &VersionSummarizedMetric{
+					Version:          v,
+					SummarizedMetric: smT,
+				})
+			}
+
+			smU, err := calculateSummarizedMetric(metrics.MetricsOverUsers)
+			if err != nil {
+				log.Logger.Debugf("unable to compute summaried metrics over users for application %s (version %d; signature %s)", application, v, *signature)
+				continue
+			}
+			entry.SummaryOverUsers = append(entry.SummaryOverUsers, &VersionSummarizedMetric{
+				Version:          v,
+				SummarizedMetric: smU,
+			})
+			result[metric] = entry
+
+			// copy data into structure for histogram calculation (to be done later)
+			// over transaction data
+			vStr := fmt.Sprintf("%d", v)
+			_, ok = byMetricOverTransactions[metric]
+			if !ok {
+				byMetricOverTransactions[metric] = make(map[string][]float64, 0)
+			}
+			(byMetricOverTransactions[metric])[vStr] = metrics.MetricsOverTransactions
+
+			_, ok = byMetricOverUsers[metric]
+			if !ok {
+				byMetricOverUsers[metric] = make(map[string][]float64, 0)
+			}
+			(byMetricOverUsers[metric])[vStr] = metrics.MetricsOverUsers
+		}
+	}
+
+	// compute histograms
+	for metric, byVersion := range byMetricOverTransactions {
+		hT, err := calculateHistogram(byVersion, 0, 0)
+		if err != nil {
+			log.Logger.Debugf("unable to compute histogram over transactions for application %s (metric %s)", application, metric)
+			continue
+		} else {
+			resultEntry := result[metric]
+			resultEntry.HistogramsOverTransactions = &hT
+			result[metric] = resultEntry
+		}
+	}
+
+	for metric, byVersion := range byMetricOverUsers {
+		hT, err := calculateHistogram(byVersion, 0, 0)
+		if err != nil {
+			log.Logger.Debugf("unable to compute histogram over users for application %s (metric %s)", application, metric)
+			continue
+		} else {
+			resultEntry := result[metric]
+			resultEntry.HistogramsOverUsers = &hT
+			result[metric] = resultEntry
+		}
+	}
+
+	// convert to JSON
+	b, err := json.MarshalIndent(result, "", "   ")
+	if err != nil {
+		http.Error(w, fmt.Sprintf("unable to create JSON response %s", string(b)), http.StatusInternalServerError)
+		return
+	}
+
+	// finally, send response
+	w.Header().Add("Content-Type", "application/json")
+	_, _ = w.Write(b)
+}
+
+// splitApplicationKey is a utility function that returns the name and namespace from a key of the form "namespace/name"
+func splitApplicationKey(applicationKey string) (string, string) {
+	var name, namespace string
+	names := strings.Split(applicationKey, "/")
+	if len(names) > 1 {
+		namespace, name = names[0], names[1]
+	} else {
+		namespace, name = "default", names[0]
+	}
+
+	return namespace, name
+}
```

Comment:
I feel like I have seen this function somewhere else in the code base? Is this a duplication?

Reply:
perhaps but it is the sort of function that we don't want to externalize. Maybe in base/util.go. I will look and see if it is duplicated.

Reply:
moved to util.SplitApplication()